File: greedy_local.cpp
#include "greedy_local.h"
#include "globals.h"

/* Constructor & Destructor */

GreedyLocal::~GreedyLocal() {
    for (int i = 0; i < n; i++)
        delete [] board[i];
    delete [] board;
}

/* Public */

void GreedyLocal::solve() {

    int tries = n / 2;
    int curr = findLowest();
    int last;

    do {
        placeLowest();
        update();

        last = curr;
        curr = findLowest();

        if (curr >= last)
            tries--;

        /* If we can't improve after N / 2 attempts, try a new random board. */
        if (!tries) {
            initialize();
            tries = n / 2;
        }

    } while (!is_solution());
}


void GreedyLocal::print() {
    /* Print the board: -1 are queens, everything else is blank */
    for (int r = 0; r < n; r++) {
        for (int c = 0; c < n; c++)
            if (board[r][c] == -1)
                cout <<"Q ";
            else
                cout << "- ";
        cout << endl;
    }
}

/* Private */

void GreedyLocal::initialize () {
    /* Create a board if it doesn't exist */
    if (!board) {
        board = new Tile*[n];
        for (int r = 0, i; r < n; r++)
            board[r] = new Tile[n];
    }

    /* Zero out old board */
    for (int r = 0; r < n; r++)
        for (int c = 0; c < n; c++)
            board[r][c] = 0;

    /* Clear the queens list */
    queens.clear();

    /* Assign queens 1 per column randomly */
    for (int r = 0, c; r < n; r++) {
        c = rand() % n;
        board[r][c] = -1;
        queens.push_back(make_pair(r, c));
    }

    update();
}


void GreedyLocal::update() {
    int count = 0;

    /* For ever row, calculate the heuristic value as if the queen was moved
       to every sqaure in that row */
    for (int r = 0; r < n; r++) {
        board[queens[r].first][queens[r].second] = 0;
        for (int c = 0; c < n; c++) {
            board[r][c] = -1;
            for (int k = 0; k < n; k++)
                if (k != r)
                    count += heuristicFunction(queens[k]);
            count += heuristicFunction(make_pair(r, c));
            board[r][c] = (queens[r].first == r && queens[r].second == c) ? -1 : count;
            count = 0;
        }
    }
}


int GreedyLocal::findLowest() {
    int min = (board[0][0] != -1) ? board[0][0] : board[0][1];
    minimums.clear();

    /* Find the minimum value on the board */
    for (int r = 0; r < n; r++)
        for (int c = 0; c < n; c++)
            if (board[r][c] < min && board[r][c] != -1)
                min = board[r][c];

    /* Find all the mimimums values */
    for (int r = 0; r < n; r++)
        for (int c = 0; c < n; c++)
            if (board[r][c] == min)
                minimums.push_back(make_pair(r, c));

    return min;
}


void GreedyLocal::placeLowest () {
    /* Pick a random minimum location */
    int move = rand() % minimums.size();
    int row = minimums[move].first;
    int col = minimums[move].second;

    /* Find the queen and replace it with a heuristic value */
    for (int c = 0; c < n; c++)
        if (board[row][c] == -1) {
            board[row][col] = -1;           // Put new queen
            queens[row].first = row;
            queens[row].second = col;

            board[row][c] = 0;              // Remove old queen
            break;
        }
}


/* Number of colliding queens */
int GreedyLocal::heuristicFunction(pair<int, int> queen) {
    int count = 0;

    /* down & right */
    for (int r = queen.first + 1, c = queen.second + 1; r < n && c < n; r++, c++)
           if (board[r][c] == -1)
               count++;

    /* down & left */
    for (int r = queen.first + 1, c = queen.second - 1; r < n && c > -1; r++, c--)
           if (board[r][c] == -1)
               count++;

    /* down */
    for (int r = queen.first + 1, c = queen.second; r < n; r++)
        if (board[r][c] == -1)
            count++;

    return count;
}


bool GreedyLocal::is_solution () {
    /* Check the heuristic for each queen */
    for (int i = 0; i < n; i++)
        if (heuristicFunction(queens[i]))
            return false;

    return true;
}

File: greedy_local.py
import random
import numpy as np

class GreedyLocal:

    def __init__(self, N):
        self.N = N

        self.initialize()

    def __str__(self):
        pass

    def solve():
        """


        """
        pass

    def initialize(self):
        """

        """
        self.board = np.array([[0 for _ in range(9)] for _ in range(9)], dtype=np.int8)
        self.queens = []
        self.minimums = []

        for r in range(self.N):
            c = random.randint(0, N - 1)
            self.board[r][c] = -1
            queens.appned(tuple(r, c))

        self.update()

    def update(self):
        """

        """
        count = 0

        for r in range(N):
            self.board[queens[r][0]][queens[r][1]] = 0
            for c in range(N):
                self.board[r][c] = -1
                for k in range(N):
                    if k != r:
                        count += self.heuristic(queens[k])

                count += self.heuristic(tuple(r, c))
                board[r][c] = -1 if (queens[r][0] == r && queens[r][1] == c) else count
                count = 0

    def find_lowest(self):
        """

        """
        min = self.board[0][0] if self.board[0][0] != -1 else self.board[0][1]
        self.minimums = []

        for r in range(self.N):
            for c in range(self.N):
                if self.board[r][c] < min and self.board[r][c] != -1:
                    min = self.board[r][c]

        for r in range(self.N):
            for c in range(self.N):
                if self.board[r][c] == min:
                    self.minimums.append(tuple(r, c))

        return min

    def place_lowest(self):
        """

        """
        move = random.randint(0, len(self.minimums) - 1)
        row, col = minimum[move][0], minimum[move][1]

        for c in range(self.N):
            if self.board[row][c] == -1:
                self.board[row][col] = -1
                queens[row][0], queens[row][1] = row, col

                self.board[row][c] = 0
                break

    def heuristic(self, queen):
        """

        """
        count = 0
        for r, c in zip(range(queen[0]+1, self.N), range(queen[1]+1, self.N))):
            if self.board[r][c] == -1:
                count += 1

        for r, c in zip(range(queen[0]+1, self.N), range(queen[1]-1, -1, -1))):
            if self.board[r][c] == -1:
                count += 1

        for r in range(queen[0]+1, self.N):
            if self.board[r][queen[1]] == -1:
                count += 1

        return count

    def is_solution(self):
        """

        """
        for i in range(self.N):
            if self.heuristic(self.queens[i]):
                return False

        return True
